File: app/src/main/java/com/kieran/winnipegbusbackend/FavouriteStopsList.java
package com.kieran.winnipegbusbackend;

import android.util.Xml;

import com.kieran.winnipegbus.Activities.BaseActivity;
import com.kieran.winnipegbusbackend.enums.FavouritesListSortType;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xmlpull.v1.XmlSerializer;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

public class FavouriteStopsList {
    private final static String FILES_DIR = BaseActivity.filesDir + "/favourites.xml";

    private final static String FAVOURITE_STOPS_TAG = "favouriteStops";
    private final static String FAVOURITE_STOP_TAG = "favouriteStop";
    private final static String STOP_NUMBER_TAG =  "stopNumber";
    private final static String STOP_NAME_TAG = "stopName";
    private final static String TIMES_USED_TAG = "timesUsed";
    private final static String ALIAS_TAG = "alias";

    private static List<FavouriteStop> favouritesList = new ArrayList<>();
    public static boolean isLoadNeeded = true;
    private final static String XMLFeature = "http://xmlpull.org/v1/doc/features.html#indent-output";

    public static void addToFavourites(FavouriteStop favouriteStop) {
        if(!contains(favouriteStop)) {
            favouritesList.add(favouriteStop);
            saveFavouriteStops();
        }
    }

    public static boolean contains(FavouriteStop favouriteStop) {
       return contains(favouriteStop.getNumber());
    }

    public static boolean contains(int stopNumber) {
        for(FavouriteStop fs : favouritesList)
            if(fs.getNumber() == stopNumber)
                return true;

        return false;
    }

    public static void remove(int stopNumber) {
        favouritesList.remove(getFavouriteStopByStopNumber(stopNumber));
        saveFavouriteStops();
    }

    public static FavouriteStop getFavouriteStopByStopNumber(int stopNumber) {
        for(FavouriteStop fs : favouritesList)
            if(fs.getNumber() == stopNumber)
                return fs;
        return null;
    }

   public static boolean loadFavourites() {
       if(favouritesList.isEmpty())
           isLoadNeeded = true;
       if(isLoadNeeded) {
           try {
               Document XMLDocument = getXML(new FileInputStream(FILES_DIR)).getResult();
               NodeList favouriteStops = XMLDocument.getElementsByTagName(FAVOURITE_STOP_TAG);

               for (int r = 0; r < favouriteStops.getLength(); r++) {
                   Node curr = favouriteStops.item(r);
                   int stopNumber = Integer.parseInt(getValue(STOP_NUMBER_TAG, curr));
                   String stopName = getValue(STOP_NAME_TAG, favouriteStops.item(r));
                   int timesUsed = Integer.parseInt(getValue(TIMES_USED_TAG, curr));
                   String alias = getValue(ALIAS_TAG, favouriteStops.item(r));

                   FavouriteStop favouriteStop = new FavouriteStop(stopName, stopNumber, timesUsed);
                    if(alias != null)
                        favouriteStop.setAlias(alias);

                   if(!contains(favouriteStop))
                        favouritesList.add(favouriteStop);
               }
               isLoadNeeded = false;
           } catch (Exception e) {
               isLoadNeeded = true;
           }
       }
        return isLoadNeeded;
    }

    public static boolean saveFavouriteStops() {
        isLoadNeeded = true;
        try {
            FileOutputStream fos;
            fos = new FileOutputStream(FILES_DIR);
            XmlSerializer serializer = Xml.newSerializer();
            serializer.setOutput(fos, "UTF-8");
            serializer.startTag("", FAVOURITE_STOPS_TAG);
            serializer.setFeature(XMLFeature, true);

            for (int i = 0; i < favouritesList.size(); i++) {
                try {
                    FavouriteStop favouriteStop = favouritesList.get(i);
                    serializer.startTag("", FAVOURITE_STOP_TAG);
                    serializer.startTag("", STOP_NUMBER_TAG);
                    serializer.text(Integer.toString(favouritesList.get(i).getNumber()));
                    serializer.endTag("", STOP_NUMBER_TAG);

                    serializer.startTag("", STOP_NAME_TAG);
                    serializer.text(favouriteStop.getName());
                    serializer.endTag("", STOP_NAME_TAG);

                    if (favouriteStop.getAlias() != null) {
                        serializer.startTag("", ALIAS_TAG);
                        serializer.text(favouriteStop.getAlias());
                        serializer.endTag("", ALIAS_TAG);
                    }

                    serializer.startTag("", TIMES_USED_TAG);
                    serializer.text(Integer.toString(favouriteStop.getTimesUsed()));
                    serializer.endTag("", TIMES_USED_TAG);

                    serializer.endTag("", FAVOURITE_STOP_TAG);
                }catch (Exception e) {
                    //intentionally blank
                }
            }
            serializer.endTag("", FAVOURITE_STOPS_TAG);
            serializer.endDocument();
            serializer.flush();
            fos.close();
            return true;
        } catch (IOException e) {
            return false;
        }
    }

    public static void sort(final FavouritesListSortType sortType) {
        Collections.sort(favouritesList, new Comparator<FavouriteStop>() {
            @Override
            public int compare(FavouriteStop stop1, FavouriteStop stop2) {
                switch (sortType) {
                    case STOP_NUMBER_ASC:
                        return stop1.getNumber() - stop2.getNumber();
                    case STOP_NUMBER_DESC:
                        return -(stop1.getNumber() - stop2.getNumber());
                    case FREQUENCY_ASC:
                        return (stop1.getTimesUsed() - stop2.getTimesUsed());
                    case FREQUENCY_DESC:
                        return -(stop1.getTimesUsed() - stop2.getTimesUsed());
                    default:
                        return stop1.getNumber() - stop2.getNumber();
                }
            }
        });
    }

    public static List<FavouriteStop> getFavouriteStopsSorted(final FavouritesListSortType sortType) {
        loadFavourites();

        sort(sortType);
        return favouritesList;
    }

    public static FavouriteStop get(int position) {
        return favouritesList.get(position);
    }

    private static String getValue(String tag, Node originalNode) {
        try {
            Node node = ((Element)originalNode).getElementsByTagName(tag).item(0).getFirstChild();
            return node.getNodeValue();
        }catch (Exception e) {
            return null;
        }
    }

    private static LoadResult<Document> getXML(InputStream inputStream)  {
        try {
            DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
            DocumentBuilder db = dbf.newDocumentBuilder();
            Document XMLDocument = db.parse(inputStream);

            return new LoadResult<>(XMLDocument, null);
        } catch (Exception e) {
            return new LoadResult<>(null, e);
        }
    }
}

File: app/src/main/java/com/kieran/winnipegbusbackend/FavouriteStopsList.kt
package com.kieran.winnipegbusbackend

import android.util.Xml

import com.kieran.winnipegbus.activities.BaseActivity
import com.kieran.winnipegbusbackend.common.FavouriteStop
import com.kieran.winnipegbusbackend.common.LoadResult
import com.kieran.winnipegbusbackend.enums.FavouritesListSortType

import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node

import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStream
import java.util.ArrayList
import java.util.Collections

import javax.xml.parsers.DocumentBuilderFactory

object FavouriteStopsList {
    private val FILES_DIR = BaseActivity.filesDirPath!! + "/favourites.xml"

    private val FAVOURITE_STOPS_TAG = "favouriteStops"
    private val FAVOURITE_STOP_TAG = "favouriteStop"
    private val STOP_NUMBER_TAG = "stopNumber"
    private val STOP_NAME_TAG = "stopName"
    private val TIMES_USED_TAG = "timesUsed"
    private val ALIAS_TAG = "alias"

    private val favouritesList = ArrayList<FavouriteStop>()
    var isLoadNeeded = true
    private val XMLFeature = "http://xmlpull.org/v1/doc/features.html#indent-output"

    fun addToFavourites(favouriteStop: FavouriteStop) {
        if (!contains(favouriteStop)) {
            favouritesList.add(favouriteStop)
            saveFavouriteStops()
        }
    }

    operator fun contains(favouriteStop: FavouriteStop): Boolean {
        return contains(favouriteStop.number)
    }

    operator fun contains(stopNumber: Int): Boolean {
        return favouritesList.any { it.number == stopNumber }
    }

    fun remove(stopNumber: Int) {
        favouritesList.remove(getFavouriteStopByStopNumber(stopNumber))
        saveFavouriteStops()
    }

    fun getFavouriteStopByStopNumber(stopNumber: Int): FavouriteStop? {
        return favouritesList.firstOrNull { it.number == stopNumber }
    }

    fun loadFavourites(): Boolean {
        if (favouritesList.isEmpty())
            isLoadNeeded = true
        if (isLoadNeeded) {
            try {
                val XMLDocument = getXML(FileInputStream(FILES_DIR)).result
                val favouriteStops = XMLDocument!!.getElementsByTagName(FAVOURITE_STOP_TAG)

                for (r in 0 until favouriteStops.length) {
                    val curr = favouriteStops.item(r)
                    val stopNumber = Integer.parseInt(getValue(STOP_NUMBER_TAG, curr))
                    val stopName = getValue(STOP_NAME_TAG, favouriteStops.item(r))
                    val timesUsed = Integer.parseInt(getValue(TIMES_USED_TAG, curr))
                    val alias = getValue(ALIAS_TAG, favouriteStops.item(r))

                    val favouriteStop = FavouriteStop(stopName!!, stopNumber, timesUsed)
                    if (alias != null)
                        favouriteStop.alias = alias

                    if (!contains(favouriteStop))
                        favouritesList.add(favouriteStop)
                }
                isLoadNeeded = false
            } catch (e: Exception) {
                isLoadNeeded = true
            }

        }
        return isLoadNeeded
    }

    fun saveFavouriteStops(): Boolean {
        isLoadNeeded = true
        try {
            val fos = FileOutputStream(FILES_DIR)
            val serializer = Xml.newSerializer()
            serializer.setOutput(fos, "UTF-8")
            serializer.startTag("", FAVOURITE_STOPS_TAG)
            serializer.setFeature(XMLFeature, true)

            for (i in favouritesList.indices) {
                try {
                    val favouriteStop = favouritesList[i]
                    serializer.startTag("", FAVOURITE_STOP_TAG)
                    serializer.startTag("", STOP_NUMBER_TAG)
                    serializer.text(Integer.toString(favouritesList[i].number))
                    serializer.endTag("", STOP_NUMBER_TAG)

                    serializer.startTag("", STOP_NAME_TAG)
                    serializer.text(favouriteStop.name)
                    serializer.endTag("", STOP_NAME_TAG)

                    if (favouriteStop.alias != null) {
                        serializer.startTag("", ALIAS_TAG)
                        serializer.text(favouriteStop.alias)
                        serializer.endTag("", ALIAS_TAG)
                    }

                    serializer.startTag("", TIMES_USED_TAG)
                    serializer.text(Integer.toString(favouriteStop.timesUsed))
                    serializer.endTag("", TIMES_USED_TAG)

                    serializer.endTag("", FAVOURITE_STOP_TAG)
                } catch (e: Exception) {
                    //intentionally blank
                }

            }
            serializer.endTag("", FAVOURITE_STOPS_TAG)
            serializer.endDocument()
            serializer.flush()
            fos.close()
            return true
        } catch (e: IOException) {
            return false
        }

    }

    fun sort(sortType: FavouritesListSortType) {
        Collections.sort(favouritesList) { stop1, stop2 ->
            when (sortType) {
                FavouritesListSortType.STOP_NUMBER_ASC -> stop1.number - stop2.number
                FavouritesListSortType.STOP_NUMBER_DESC -> -(stop1.number - stop2.number)
                FavouritesListSortType.FREQUENCY_ASC -> stop1.timesUsed - stop2.timesUsed
                FavouritesListSortType.FREQUENCY_DESC -> -(stop1.timesUsed - stop2.timesUsed)
            }
        }
    }

    fun getFavouriteStopsSorted(sortType: FavouritesListSortType): List<FavouriteStop> {
        loadFavourites()

        sort(sortType)
        return favouritesList
    }

    operator fun get(position: Int): FavouriteStop {
        return favouritesList[position]
    }

    private fun getValue(tag: String, originalNode: Node): String? {
        return try {
            val node = (originalNode as Element).getElementsByTagName(tag).item(0).firstChild
            node.nodeValue
        } catch (e: Exception) {
            null
        }

    }

    private fun getXML(inputStream: InputStream): LoadResult<Document> {
        return try {
            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            val XMLDocument = db.parse(inputStream)

            LoadResult(XMLDocument, null)
        } catch (e: Exception) {
            LoadResult<Document>(null, e)
        }

    }
}
